src/MassTransit/Pipeline/ConsumerFactories/BatchConsumerFactory.cs

```diff
@@ -65,18 +84,67 @@ void IProbeSite.Probe(ProbeContext context)
 
         async Task<BatchConsumer<TConsumer, TMessage>> Add(ConsumeContext<TMessage> context)
         {
-            if (_currentConsumer != null)
+            var groupKey = GetGroupKey(context);
+            BatchConsumer<TConsumer, TMessage> consumer = _currentConsumers.Get(groupKey);
+
+            if (consumer != null)
             {
                 if (context.GetRetryAttempt() > 0)
-                    await _currentConsumer.ForceComplete().ConfigureAwait(false);
+                    await consumer.ForceComplete().ConfigureAwait(false);
+            }
+
+            if (consumer == null || consumer.IsCompleted)
+            {
+                consumer = new BatchConsumer<TConsumer, TMessage>(_messageLimit, _timeLimit, _collector, _dispatcher, _consumerFactory, _consumerPipe);
+                _currentConsumers.Set(groupKey, consumer);
+            }
+
+            await consumer.Add(context).ConfigureAwait(false);
+
+            return consumer;
+        }
+
+        object GetGroupKey(ConsumeContext context)
```

Comment:
why do you need function if you have `Func`?

Reply:
just to avoid duplicating the null check :)

Reply:
`functor?.Invoke()` as easy as C#